### ISSUE TITLE
[DISCUSSION] Dynamic loading style (CSS) and theming support

### ISSUE THREAD

**Issue by fpliger** [pyscript/pyscript#340]
PyScript currently "requires" users to load the CSS explicitly and separately from the JS. This has a few limitations:

* locks the flexibility of customizing style
* forces to load all the styles before the JS logic itself (adding up to the loading time, even if just a little)
* not friendly to support theming 
* users have to specify 2 cmds to load PyScript on their pages (link to CSS and script to js)

If we switch to lading CSS dynamically within the scope of pyscript.js we can:

* have a default global style that loads if not stylesheet or theme has been specified
* allow users to create stylesheets/themes on their own that can be used instead of the default one
* have more control over what's being loaded into the page and when
* reduce the amounts of steps needed for a user to start using PyScript

In terms of implementation, we could allow users to specify a theme for the app in the AppConfig yaml that can be loaded in runtime. 

Would love to hear opinions about the good and bad of the different approaches and possible alternative solutions. :)

**Comment by imjul1an** [comment]
Framework-agnostic CSS-in-JS libraries (see [JSS](https://cssinjs.org/)) can help us to load CSS dynamically and address some of the concerns you've mentioned above.

imho. for theming configuration, I'd suggest sticking to JS "world" practices and allowing defining themes in *.js, rather than in yaml. JSS should already address that as well.

**Comment by fpliger** [comment]
Thanks for the suggestion. Yeah, sticking with best practices as much as possible is definitely a good principle we should try to stick with. 

From your comment, I also realized that my proposal wasn't well explained on my part. I'm not suggesting that we add the style rules and all in the yaml itself but use the yaml to provide pyscript the metadata to access the theme somewhere (so it can also load its related assets), with a name/location/url.

**Comment by verhulstm** [comment]
i would like to support simple pure css themes. i do also like the idea of supporting CSS-in-JS like JSS. so i would like both.

many beginners will want to dig into learning python and web tech, and making or modifying a theme is a good starting point for a ton of folks. JSS is too complex for some who are just starting out in their learning. for them, i want to be able to point to a simple pure css theme as a starting point. and then they can move on to JSS

**Comment by imjul1an** [comment]
@verhulstm agree. For that, we could provide a user-friendly API to do theming. We can still leverage JSS-like frameworks and wrap it with a simple API. Here is a simple example of what it might look like.

We introduce a provider component that would wrap the other components and propagate styles/themes to other components used in the Pyscript.

We expose the configuration that is easy to reason about for novice users, as well as support more complex configurations for more advanced users.

```js
const customTheme = extendTheme({
  semanticTokens: {
    colors: {
      error: 'red.500',
      success: 'green.500',
      primary: {
        default: 'red.500',
        _dark: 'red.400',
      },
      secondary: {
        default: 'red.800',
        _dark: 'red.700',
      },
    },
    'py-button': {
      background-color: 'black.500',
      color: 'white.200'
    },
   'py-title': {
     color: 'yellow.500',
     fontSize: '14px',
   },
   'py-input-box': {
     color: 'black.200',
     font-size: '14px'
   }
  },
})
```

```html
<py-provider theme={customTheme}>
  ....
  <py-title></py-title>
  <py-inputbox><py-inputbox>
  <py-button></py-button>
<py-provider/>
```

**Comment by verhulstm** [comment]
can we support all the above in a css file?

**Comment by pauleveritt** [comment]
+1 for pure CSS as an option. Peter quite rightly bangs the drum about entry, that `<script type="text/python">` is too much typing vs. `<py-script>`. Same applies to "you have to do npm install and learn a JSS framework, produce a build, etc.".

Frankly, I hope PyScript splits and doesn't mandate any CSS, as you won't choose what I want (Bulma). Something built atop should be opinionated. My two cents.

**Comment by fpliger** [comment]
This is a great discussion. 

I get the point about not mandating CSS and keeping PyScript as lean as possible, otoh, if we really want to target non-technical users to empower them with easy to build UI capabilities, we do need to have an opinonated options for them to just drop new components on a page "that just works".

With that said, I think this can be optional. Either by loading the theme (CSS) as an option or vice versa (basically opt-in vs opt-out). Or even splitting theming from PyScript core and let it be a separate load (this would mean that users need to do extra work though, that is harder for non-technical users).

**Comment by fpliger** [comment]
That sounds great. Yeah, demos are more effective than words, for sure! 😄

**Comment by marimeireles** [comment]
I guess this is very related to this issue: https://github.com/pyscript/pyscript/issues/78
I'm gonna centralize the discussion here, but ^ 78 is relevant for it :)
